### DISCUSSION TITLE
APP_KTL_VERSIONS in App Settings table - update automatically?

### DISCUSSION THREAD

**Discussion by IntegriTivity** [cortexrd/Knack-Toolkit-Library#331] [Q&A] [answered]
Should the value for APP_KTL_VERSIONS in App Settings table update automatically when I save my app version in the Javascript? 
Because it doesn't, so after I installed the KTL, my users got the flashing update message.  When I updated the table value to match the version in my Heartbeat, the flashing message went away.  Am I supposed to change the value manually before doing a BROADCAST NOW?  If it should be automatic, can you give me any clues what I may have missed in the configuration?

**Comment by cortexrd** [comment]
Hi @IntegriTivity,

I had this question in the past.  You're not the first who's experienced that issue.

The setup is supposed to be automatic.  When the iFrameWnd is initialized the first time, that APP_KTL_VERSIONS entry should be created so you don't have to worry about it.

I'll investigate that.  

In the meantime, is it ok now or do you need any urgent assistance?

Norm

**Reply by IntegriTivity** [reply]
Not urgent as flashing is gone. I renamed the form to SW Update and now if I change my version in JS, my Heartbeat notices it and tells me to update. So up in the SW Update form, the version up there is not supposed to be the new one, it's showing the "old one". Okay, I just was expecting that the Broadcast Now link was going to broadcast the version shown there, since I read left to right. But now I understand that is not the case and when I clicked the link, THEN it update the value to be the new version.  I am using KTL 0.27.10 and I do have an APP version specified (3.0.3).

**Reply by cortexrd** [reply]
>I renamed the form to SW Update

You probably mean the **grid** (not form)

You got it right!  

Let's review the process:

1. In the SW Update grid, the versions shown (APP+KTL) are the ones that all users are currently running.
2. When you click BROWADCAST NOW, an API call updates that grid value.
3. Within 20 seconds, all users' iFrameWnd will see that new version and since it doesn't match their version at top right (currently running), it will force their browser to refresh itself, fetching all the latest Builder structure and code.

The magic happens in the point number 4 of this settings' page:  https://github.com/cortexrd/Knack-Toolkit-Library/wiki/Advanced-Features#iframewnd-page


4. On a third row, add a Grid view:
- Displays App Settings
- Title: App Settings _ar=20

It's the auto refresh = 20 seconds that does the job, via the hidden code behind it.

Hope this clarifies everything.  

I will do a video about this and improve the documentation.

Norm

**Comment by cortexrd** [comment]
To answer your question...

>Should the value for APP_KTL_VERSIONS in App Settings table update automatically when I save my app version in the Javascript?

No, saving the version in the Javascript does not update the record.  And this is by design.

Actually, this is the trick I use to gain full control over precisely WHEN the update is pushed.

When these two values are different, a forced refresh is triggered in all opened browsers.

This is true except for yours; as having the Developer role, you get the yellow warning that you must do the Broadcast now. This is what will stop the refresh loop they all had.

What I don't understand is that there's code to kick start the process for the first time, i.e. when that record does not exist.  The KTL creates it via API.  But for some reason, it doesn't always work.

So I'm trying to pinpoint that scenario / setup / sequence.

If you have anything to share, it would help.

What KTL version are you using in the Loader?  Do you have a blank or a specific version between the two single quotes?

Norm

**Reply by IntegriTivity** [reply]
So then in my IframeWnd, when is the AppSettings grid supposed to show the new version that I broadcast?

**Reply by cortexrd** [reply]
See my reply above, about **point number 4**.

**Comment by CSWinnall** [comment]
Hi Norm

We had this issue too.  We had to manually update the version otherwise
everyone had a forced refresh loop.  We have not been able to resolve this
issue either.

Craig

**Comment by CSWinnall** [comment]
Also using it with the Dev environment means we constantly get that message
flashing up that we aren't up to date.

**Reply by cortexrd** [reply]
1. Craig, can you reproduce that behavior at will?
2. Do you have a sequence for me?

I may have to connect remotely to see what's happening, because I can't reproduce the bug in any pf my apps.

Norm

**Reply by CSWinnall** [reply]
No I can't reproduce the issue with the dev environment any more. I'll keep you posted if it happens again.

**Reply by CSWinnall** [reply]
The version number not changing is an issue though.
1. I update the app version number in JS builder pane
2. I Go back to app refresh I get the message saying the app is not up to date please broadcast now
3. I go to sw update then click broadcast now - Nothing happens to my page.  I do get the message saying updating app. Also the version Number in the sw update page stays at the previous number.
5. I refresh the browser and I keep getting the message saying you need to broadcast an update now. This also causes a continuous loop for other users. This message will not go away until I update the app version number manually in the SW Update grid. Then everything is back to normal for me and The other users get updated.

**Reply by cortexrd** [reply]
When you click the BROADCAST NOW button, can you breakpoint on this line and check the parameters, to see if they're all ok?

`ktl.core.knAPI(cfg.appBcstSWUpdateViewId, data[0].id, apiData, 'PUT', [cfg.appBcstSWUpdateViewId])`

Note that all titles and labels are strict and are case sensitive, including the tables.

Norm

**Reply by CSWinnall** [reply]
Hi Norm

This is my setup:
![image](https://github.com/user-attachments/assets/3b137c10-1153-4015-909f-aaa39b0097ce)

This is where I added breakpoints.
![image](https://github.com/user-attachments/assets/c1cb0d64-4371-4d45-b921-c7f70a795435)

cfg.appBcstSWUpdateViewId is undefined.

Craig

**Reply by cortexrd** [reply, marked as answer]
Rename your view from **App Settings** to **SW Update**.

FYI, this is the line that sets this variable from the title:

`appBcstSWUpdateViewId: ktl.core.getViewIdByTitle('SW Update', Knack.router.current_scene_key, true),`

**Reply by CSWinnall** [reply]
Great Thanks

**Reply by CSWinnall** [reply]
Sorry about that

**Reply by cortexrd** [reply]
No problem!  It happens to all of us once in a while :)

I wonder why Knack don't make it so that the **view title** matches the **view name** by default.  It's so annoying to have to change them all every time.

**Reply by IntegriTivity** [reply]
Thanks, renaming my view from to SW Update did the trick and then I added your explanatin of how it works to the description of my grid for my future reference.

**Reply by cortexrd** [reply]
Glad you made it work!  I will update the documentation to make this a bit clearer.

**Comment by CSWinnall** [comment]
Yes of course I'll set it up today and send you thevlink
Kind regards,

Craig Winnall
07543423538
